test: fix issues with runtime/connectivity.go test

ianvernon opened the pull request:
The test `RuntimeValidatedConntrackTest` was not properly testing the Conntrack
and ConntrackLocal endpoint configuration options. To start, it was testing
Cilium in its default policy enforcement mode, which by default allows all
traffic for endpoints which no rules select in the policy repository. This is an
issue because the test was ran with no policy imported, meaning that the
configuration flags were not really being tested appropriately.

The following changes were made to fix this, and to  ensure the test has true
parity with its Bash-script counterpart, `tests/01-ct.sh`:

* Import the same policy from the corresponding Bash test.
* Set policy enforcement mode for Cilium to be 'always'.
* Add checks for curl, and bidirectional connectivity.

The following changes were made to enhance the test:

* Miscellaneous variable renaming to make it easier to read the test code.
* Combining all Its into one It, which results in the containers used for the
tests only having to be launched and removed once, instead of three times (one
for each previous It).
* Directly porting `connectivity_test` from the Bash test to Ginkgo.

Also made logs in `test/helpers/cilium.go:PolicyImportAndWait` use logrus
Fields to reduce the amount of custom string formatting in its log messages.

Signed-off by: Ian Vernon <ian@cilium.io>"


Fixes: #2835
Related-to: #1841 

```release-note
Fix issues with endpoint Conntrack configuration tests
```

**How to test**:
```ginkgo --focus="RuntimeValidatedConntrackTest" -v```

Commit:
test: fix issues with runtime/connectivity.go test

The test `RuntimeValidatedConntrackTest` was not properly testing the Conntrack
and ConntrackLocal endpoint configuration options. To start, it was testing
Cilium in its default policy enforcement mode, which by default allows all
traffic for endpoints which no rules select in the policy repository. This is an
issue because the test was ran with no policy imported, meaning that the
configuration flags were not really being tested appropriately.

The following changes were made to fix this, and to  ensure the test has true
parity with its Bash-script counterpart, `tests/01-ct.sh`:

* Import the same policy from the corresponding Bash test.
* Set policy enforcement mode for Cilium to be 'always'.
* Add checks for curl, and bidirectional connectivity.

The following changes were made to enhance the test:

* Miscellaneous variable renaming to make it easier to read the test code.
* Combining all Its into one It, which results in the containers used for the
tests only having to be launched and removed once, instead of three times (one
for each previous It).

Also made logs in `test/helpers/cilium.go:PolicyImportAndWait` use logrus
Fields to reduce the amount of custom string formatting in its log messages.

Signed-off by: Ian Vernon <ian@cilium.io>"

aanm commented on test/runtime/connectivity.go:
```diff
@@ -280,135 +280,200 @@ var _ = Describe("RuntimeValidatedConntrackTest", func() {
+		for _, conntrackOptionMode := range conntrackOptionModes {
+			By(fmt.Sprintf("Testing ConntrackLocal endpoint configuration option: %s", conntrackOptionMode))
+			status := vm.EndpointSetConfig(endpoints[helpers.Server], helpers.OptionConntrackLocal, conntrackOptionMode)
+			Expect(status).Should(BeTrue(), fmt.Sprintf("could not set %s=%s on endpoint %q", helpers.OptionConntrackLocal, helpers.OptionDisabled, helpers.Server))
```
Please test if the output of `cilium endpoint config <ID>` returns `ConntrackLocal Enabled`

ianvernon replied:
Added a function that sets an option and then checks it for the provided endpoint for all cases in this test.

aanm commented on test/runtime/connectivity.go:
```diff
@@ -280,135 +280,200 @@ var _ = Describe("RuntimeValidatedConntrackTest", func() {
 
-		status := vm.EndpointSetConfig(endpoints[helpers.Server], helpers.OptionConntrackLocal, helpers.OptionEnabled)
-		Expect(status).Should(BeTrue(), fmt.Sprintf("could not set %s=%s on endpoint %q", helpers.OptionConntrackLocal, helpers.OptionEnabled, helpers.Server))
+		status := vm.EndpointSetConfig(endpoints[helpers.Server], helpers.OptionConntrack, helpers.OptionDisabled)
```
Same here

ianvernon replied:
Added a function that sets an option and then checks it for the provided endpoint for all cases in this test.

aanm commented on test/runtime/manifests/ct-test-policy.json:
```diff
@@ -0,0 +1,51 @@
+[{
```
Where can we add a description about the behavior this policy offers without reading it through all of it? (Not only for this one but for all present in `manifests` directory)

ianvernon replied:
Shouldn't the policy language itself be enough? I think it's fair to assume a certain level of knowledge of the writers / users of tests.

aanm replied:
I mean, it would be useful to get the overall idea of what the test is. The name "ct-test-policy" is not pretty descriptive to me. I was just thinking if we had policies that could be used in multiple tests. Let's imagine there was a policy called "allow client to talk with server on port 80" that could be a policy that could be used in multiple tests for example.

ianvernon replied:
We could do this by splitting up the policy into separate files with more descriptive names? Right now, at least with this test, it's somewhat of a hodgepodge of policy rules that make it hard to specify succinctly in the filename itself what it does.

Commit:
api: add endpoint configuration as part of endpoint model

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: get configuration model in endpoint model

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/helpers: add GetEndpointMutableConfigurationOption function

Add helper function which gets the value of a configuration option for a specific endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test/runtime: add checks for setting of configuration values

Check that values that are set via  are indeed set for each endpoint. Add a
variety of helper functions / refactoring to reduce code duplication.

Signed-off by: Ian Vernon <ian@cilium.io>
